### PR TITLE
Sox

### PR DESCRIPTION
**Pull request by jcftang**
This PR add the sox sound processing application to spack, it also adds a number of audio codecs that I needed for sox.

### CODE REVIEW

File: var/spack/repos/builtin/packages/id3lib/package.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class Id3lib(AutotoolsPackage):
+    """Library for manipulating ID3v1 and ID3v2 tags"""
+
+    homepage = "http://id3lib.sourceforge.net/"
+    url      = "https://downloads.sourceforge.net/project/id3lib/id3lib/3.8.3/id3lib-3.8.3.tar.gz"
+
+    version('3.8.3', '19f27ddd2dda4b2d26a559a4f0f402a7')
+
+    def url_for_version(self, version):
+        return 'https://downloads.sourceforge.net/project/id3lib/id3lib/{0}/id3lib-{0}.tar.gz'.format(version)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/id3lib/package.py]: This shouldn't be necessary. Spack has no trouble parsing this URL in the latest develop branch:
```
$ spack url parse https://downloads.sourceforge.net/project/id3lib/id3lib/3.8.3/id3lib-3.8.3.tar.gz
==> Parsing URL: https://downloads.sourceforge.net/project/id3lib/id3lib/3.8.3/id3lib-3.8.3.tar.gz

==> Matched version regex  6: r'^[a-zA-Z\\d+-]+-v?(\\d[\\da-zA-Z.]*)$'
==> Matched  name   regex  7: r'^([A-Za-z\\d+\\._-]+)$'

==> Detected:
    https://downloads.sourceforge.net/project/id3lib/id3lib/3.8.3/id3lib-3.8.3.tar.gz
                                                                  ------ ~~~~~
    name:    id3lib
    version: 3.8.3

==> Substituting version 9.9.9b:
    https://downloads.sourceforge.net/project/id3lib/id3lib/9.9.9b/id3lib-9.9.9b.tar.gz
                                                                   ------ ~~~~~~
```

File: var/spack/repos/builtin/packages/id3lib/package.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    patch("id3lib-3.8.3_gcc4.diff")
+
+    def configure_args(self):
+        args = ["--enable-debug=no"]
+        return args
+
+    def install(self, spec, prefix):
+        os.environ["LDFLAGS"] = " -lstdc++"
+        make()
+        make("install")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/id3lib/package.py]: `AutotoolsPackage` has 3 (technically 4) build phases: configure, build, and install. By default, the build phase runs `make()`, so by adding it to your install phase, you're actually running `make()` twice. If all you need to do is set an environment variable, I would recommend using `setup_environment` instead:
```python
def setup_environment(self, spack_env, run_env):
    spack_env.set('LDFLAGS', '-lstdc++')
```

File: var/spack/repos/builtin/packages/sox/package.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+class Sox(AutotoolsPackage):
+    """SoX, the Swiss Army knife of sound processing programs."""
+
+    homepage = "http://sox.sourceforge.net/Main/HomePage"
+    url      = "https://downloads.sourceforge.net/project/sox/sox/14.4.2/sox-14.4.2.tar.bz2"
+
+    version('14.4.2', 'ba804bb1ce5c71dd484a102a5b27d0dd')
+
+    def url_for_version(self, version):
+        return 'https://downloads.sourceforge.net/project/sox/sox/{0}/sox-{0}.tar.bz2'.format(version)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sox/package.py]: You shouldn't need `url_for_version` here either:
```
$ spack url parse https://downloads.sourceforge.net/project/sox/sox/14.4.2/sox-14.4.2.tar.bz2
==> Parsing URL: https://downloads.sourceforge.net/project/sox/sox/14.4.2/sox-14.4.2.tar.bz2

==> Matched version regex  0: r'^[a-zA-Z+._-]+[._-]v?(\\d[\\d._-]*)$'
==> Matched  name   regex  7: r'^([A-Za-z\\d+\\._-]+)$'

==> Detected:
    https://downloads.sourceforge.net/project/sox/sox/14.4.2/sox-14.4.2.tar.bz2
                                                             --- ~~~~~~
    name:    sox
    version: 14.4.2

==> Substituting version 9.9.9b:
    https://downloads.sourceforge.net/project/sox/sox/9.9.9b/sox-9.9.9b.tar.bz2
                                                             --- ~~~~~~
```

**Comment by jcftang** [comment]
@adamjstewart I've made the suggested changes

File: var/spack/repos/builtin/packages/id3lib/package.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    version('3.8.3', '19f27ddd2dda4b2d26a559a4f0f402a7')
+
+    depends_on('zlib')
+
+    # http://connie.slackware.com/~alien/slackbuilds/id3lib/build/id3lib-3.8.3_gcc4.diff
+    # this is due to some changes in the c++ standard library headers
+    patch("id3lib-3.8.3_gcc4.diff")
+
+    def configure_args(self):
+        args = ["--enable-debug=no"]
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/id3lib/package.py]: Why this?

File: var/spack/repos/builtin/packages/id3lib/id3lib-3.8.3_gcc4.diff [hunk 5]
```diff
@@ -0,0 +1,102 @@
+Submitted By: Ken Moffat <ken at linuxfromscratch dot org>
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/id3lib/id3lib-3.8.3_gcc4.diff]: Wow... It looks like the ID3LIB source code hasn't changed since 2003, and so this patch is still necessary!

File: var/spack/repos/builtin/packages/id3lib/package.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        return args
+
+    def setup_environment(self, spack_env, run_env):
+        spack_env.set('LDFLAGS', '-lstdc++')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/id3lib/package.py]: Why is this needed?

File: var/spack/repos/builtin/packages/id3lib/package.py (hunk 4 above)
Comment (jcftang) [inline var/spack/repos/builtin/packages/id3lib/package.py]: I took the same configure args and LDFLAGS as the rpmforge rpm spec file does it, I figured since they haven't changed it in years, there must be a reason to keep it like that. I don't have a good argument for having this or the LDFLAGS settings below. If there are significant objections to it, I can remove it.

Comment (citibeth) [inline var/spack/repos/builtin/packages/id3lib/package.py]: It looks like `--enable-debug=yes` doesn't work (see bug reports from 2004).  But can you try removing `setup_environment()` and see if it builds?  Adding `-lstdc++` should not be necessary, and it could be incompatible with certain C++ compilers.

Comment (jcftang) [inline var/spack/repos/builtin/packages/id3lib/package.py]: I removed those chunks of config and rebuilt sox/flac it seems to work for me
